File: tools/audit_career_layout.py
from pathlib import Path
import json, subprocess, re, os
from playwright.sync_api import sync_playwright
ROOT=Path(__file__).resolve().parents[1]
CSS=(ROOT/'src/styles.css').read_text(encoding='utf-8')
cat=(ROOT/'src/i18n/catalogs.js').read_text(encoding='utf-8').replace('export const CATALOGS','const CATALOGS')
idx=(ROOT/'src/i18n/index.js').read_text(encoding='utf-8')
idx=re.sub(r"^import .*?;\n",'',idx,count=1).replace('export ','')
I18N=cat+'\n'+idx
exported=subprocess.run(['node','tools/export-audit-screens.mjs'],cwd=ROOT,capture_output=True,text=True,check=True)
SCREENS=json.loads(exported.stdout)['screens']
outdir=ROOT/'docs/screenshots-v2.0.0';outdir.mkdir(exist_ok=True)
profiles=[
 ('pt-BR','pt-BR-career-journey','Reputação do hospital',568,320),
 ('pt-BR','pt-BR-career-journey','Mentoria',640,360),
 ('en','en-career-journey','Hospital reputation',844,390),
 ('es','es-career-journey','Reputación del hospital',932,430),
 ('pt-BR','pt-BR-career-journey','Especializações',768,1024),
 ('en','en-career-journey','Specializations',1024,768),
 ('es','es-career-journey','Especializaciones',1366,768),
 ('pt-BR','pt-BR-career-journey','Histórico recente',1920,1080),
 ('pt-BR','pt-BR-career-overview','CARREIRA E RESIDÊNCIA',360,800),
 ('pt-BR','pt-BR-career-residency','Residência',360,800),
 ('pt-BR','pt-BR-career-calendar','Agenda da semana',360,800),
 ('en','en-career-hospital','CAREER AND RESIDENCY',360,800),
 ('en','en-career-overview','Promotion progress',1366,768),
 ('es','es-career-exams','CARRERA Y RESIDENCIA',360,800),
 ('es','es-career-hospital','Rotación actual',768,1024),
]
results=[]
browser_candidates=[
 os.environ.get('MEDSIM_BROWSER'),
 r'C:\Program Files (x86)\Microsoft\Edge\Application\msedge.exe',
 r'C:\Program Files\Microsoft\Edge\Application\msedge.exe',
 r'C:\Program Files\Google\Chrome\Application\chrome.exe',
 '/usr/bin/chromium',
]
browser_path=next((candidate for candidate in browser_candidates if candidate and Path(candidate).exists()),None)
with sync_playwright() as p:
  launch_options={'headless':True,'args':['--no-sandbox','--disable-dev-shm-usage','--disable-gpu']}
  if browser_path:launch_options['executable_path']=browser_path
  browser=p.chromium.launch(**launch_options)
  for locale,screen,marker,width,height in profiles:
    context=browser.new_context(viewport={'width':width,'height':height},has_touch=width<=768,is_mobile=width<=768)
    page=context.new_page();errors=[];page.on('pageerror',lambda error:errors.append(str(error)))
    html=f'''<!doctype html><html lang="{locale}" class="touch-scroll-enabled"><head><meta name="viewport" content="width=device-width,initial-scale=1,viewport-fit=cover"></head><body class="touch-scroll-enabled"><div id="a11y-live" class="sr-only"></div><div id="a11y-alert" class="sr-only"></div>{SCREENS[screen]}</body></html>'''
    page.set_content(html,wait_until='domcontentloaded');page.add_style_tag(content=CSS);page.add_script_tag(content=I18N)
    page.evaluate('''(locale)=>{applyDocumentLocale(locale);localizeDOM(document.body,locale);}''',locale)
    text=page.locator('body').inner_text()
    metrics=page.evaluate('''()=>{const controls=[...document.querySelectorAll('button,select,input')].filter(x=>{const r=x.getBoundingClientRect();return r.width>0&&r.height>0});return{vw:innerWidth,w:document.documentElement.scrollWidth,h:document.documentElement.scrollHeight,controls:controls.length,minControl:controls.reduce((m,x)=>Math.min(m,x.getBoundingClientRect().height),999),touch:getComputedStyle(document.body).touchAction};}''')
    scroll=None
    if width==360 and metrics['h']>height+100:
      before=page.evaluate('scrollY');session=context.new_cdp_session(page)
      session.send('Input.dispatchTouchEvent',{'type':'touchStart','touchPoints':[{'x':180,'y':650,'radiusX':5,'radiusY':5,'force':1}]})
      for y in [610,560,510,460,410,360,310,260,210,160,110]:
        session.send('Input.dispatchTouchEvent',{'type':'touchMove','touchPoints':[{'x':180,'y':y,'radiusX':5,'radiusY':5,'force':1}]});page.wait_for_timeout(20)
      session.send('Input.dispatchTouchEvent',{'type':'touchEnd','touchPoints':[]});page.wait_for_timeout(250)
      after=page.evaluate('scrollY');scroll={'before':before,'after':after,'moved':after>before+20}
    safe=screen.replace('pt-BR-','').replace('en-','').replace('es-','')
    name=f'career-{locale}-{safe}-{width}x{height}.png';page.screenshot(path=str(outdir/name),full_page=True)
    results.append({'locale':locale,'screen':screen,'viewport':[width,height],'markerFound':marker in text,'overflowX':metrics['w']>metrics['vw']+1,'metrics':metrics,'scroll':scroll,'pageErrors':errors,'screenshot':f'docs/screenshots-v2.0.0/{name}'})
    context.close()
  browser.close()
fail=[]
for r in results:
  if not r['markerFound']:fail.append(f"{r['screen']}: marker")
  if r['overflowX']:fail.append(f"{r['screen']}: overflow")
  if r['metrics']['minControl']<47.5:fail.append(f"{r['screen']}: control {r['metrics']['minControl']}")
  if r['scroll'] and not r['scroll']['moved']:fail.append(f"{r['screen']}: touch scroll")
  if r['pageErrors']:fail.append(f"{r['screen']}: errors")
out={'ok':not fail,'method':'Production career DOM + production CSS + live i18n in isolated Chromium/Edge; CDP touch swipe on mobile','browser':browser_path or 'Playwright bundled Chromium','failures':fail,'profiles':results}
(ROOT/'docs/audit-career-layout-v2.0.0.json').write_text(json.dumps(out,ensure_ascii=False,indent=2)+'\n',encoding='utf-8')
print(json.dumps(out,ensure_ascii=False,indent=2))
if fail:raise SystemExit(1)

File: tools/audit-career-layout.mjs
import fs from 'node:fs';
import path from 'node:path';
import {execFileSync} from 'node:child_process';
import {createRequire} from 'node:module';
import {fileURLToPath} from 'node:url';

const root=fileURLToPath(new URL('..',import.meta.url));
const dependencyRoot='C:/Users/jonat/.cache/codex-runtimes/codex-primary-runtime/dependencies/node';
const require=createRequire(path.join(dependencyRoot,'package.json'));
const {chromium}=require('playwright');
const browserCandidates=[
  process.env.MEDSIM_BROWSER,
  'C:/Program Files (x86)/Microsoft/Edge/Application/msedge.exe',
  'C:/Program Files/Microsoft/Edge/Application/msedge.exe',
  'C:/Program Files/Google/Chrome/Application/chrome.exe',
  '/usr/bin/chromium'
].filter(Boolean);
const executablePath=browserCandidates.find(candidate=>fs.existsSync(candidate));
const exported=JSON.parse(execFileSync(process.execPath,[path.join(root,'tools/export-audit-screens.mjs')],{cwd:root,encoding:'utf8',maxBuffer:40*1024*1024}));
const css=fs.readFileSync(path.join(root,'src/styles.css'),'utf8');
const outdir=path.join(root,'docs/screenshots-v2.0.0');
fs.mkdirSync(outdir,{recursive:true});
const profiles=[
  ['pt-BR',568,320,'Reputação do hospital'],
  ['pt-BR',640,360,'Mentoria'],
  ['pt-BR',667,375,'Especializações'],
  ['en',740,360,'Hospital reputation'],
  ['es',780,360,'Reputación del hospital'],
  ['pt-BR',812,375,'Histórico recente'],
  ['en',844,390,'Hospital reputation'],
  ['es',852,393,'Mentoría'],
  ['pt-BR',896,414,'Especializações'],
  ['en',915,412,'Specializations'],
  ['es',932,430,'Reputación del hospital'],
  ['pt-BR',960,432,'Reputação do hospital'],
  ['pt-BR',768,1024,'Especializações'],
  ['en',1024,600,'Hospital reputation'],
  ['en',1024,768,'Specializations'],
  ['es',1180,820,'Especializaciones'],
  ['pt-BR',1280,720,'Histórico recente'],
  ['en',1280,800,'Recent history'],
  ['es',1366,768,'Especializaciones'],
  ['pt-BR',1440,900,'Reputação do hospital'],
  ['en',1600,900,'Mentoring'],
  ['pt-BR',1920,1080,'Histórico recente']
  ,['es',2560,1440,'Historial reciente']
];
const captureSet=new Set(['568x320','640x360','844x390','932x430','768x1024','1024x768','1366x768','1920x1080']);
const browser=await chromium.launch({headless:true,executablePath,args:['--no-sandbox','--disable-gpu']});
const results=[];
for(const [locale,width,height,marker] of profiles){
  const context=await browser.newContext({viewport:{width,height},hasTouch:width<=932,isMobile:false});
  const page=await context.newPage();
  const errors=[];
  page.on('pageerror',error=>errors.push(String(error)));
  const screen=`${locale}-career-journey`;
  await page.setContent(`<!doctype html><html lang="${locale}" class="touch-scroll-enabled"><head><meta name="viewport" content="width=device-width,initial-scale=1,viewport-fit=cover"><style>${css}</style></head><body class="touch-scroll-enabled">${exported.screens[screen]}</body></html>`,{waitUntil:'domcontentloaded'});
  const text=await page.locator('body').innerText();
  const metrics=await page.evaluate(()=>{
    const controls=[...document.querySelectorAll('button,select,input')].filter(node=>{const rect=node.getBoundingClientRect();const style=getComputedStyle(node);return rect.width>0&&rect.height>0&&style.display!=='none'&&style.visibility!=='hidden';});
    return {viewportWidth:innerWidth,documentWidth:document.documentElement.scrollWidth,documentHeight:document.documentElement.scrollHeight,visibleControls:controls.length,minControlHeight:controls.length?Math.min(...controls.map(node=>node.getBoundingClientRect().height)):null};
  });
  const name=`career-journey-${locale}-${width}x${height}.png`;
  const screenshot=captureSet.has(`${width}x${height}`)?`docs/screenshots-v2.0.0/${name}`:null;
  if(screenshot)await page.screenshot({path:path.join(root,screenshot),fullPage:true});
  results.push({locale,viewport:[width,height],markerFound:text.includes(marker),overflowX:metrics.documentWidth>metrics.viewportWidth+1,metrics,pageErrors:errors,screenshot});
  await context.close();
}
await browser.close();
const failures=[];
for(const result of results){
  if(!result.markerFound)failures.push(`${result.locale} ${result.viewport.join('x')}: marker`);
  if(result.overflowX)failures.push(`${result.locale} ${result.viewport.join('x')}: horizontal overflow`);
  if(result.metrics.minControlHeight!==null&&result.metrics.minControlHeight<43.5)failures.push(`${result.locale} ${result.viewport.join('x')}: control ${result.metrics.minControlHeight}`);
  if(result.pageErrors.length)failures.push(`${result.locale} ${result.viewport.join('x')}: page error`);
}
const report={ok:failures.length===0,method:'Production career DOM and CSS rendered in installed Chromium browser',browser:executablePath||'bundled',minimumControlHeight:44,failures,profiles:results};
fs.writeFileSync(path.join(root,'docs/audit-career-layout-v2.0.0.json'),`${JSON.stringify(report,null,2)}\n`);
console.log(JSON.stringify(report,null,2));
if(failures.length)process.exitCode=1;
